1201 failure aware scheduling

zapper59 opened the pull request:
#1201

dperny commented:
Lol wow this is a lot of commits. Is this ready for review? Can you add an explanation of what you're doing here? 

Welcome to swarmkit new friends! Feel free to reach out to me directly on the Docker community slack or through any other channel if you'd like help with contributing.

zapper59 commented:
@dperny  Hi!  everything in this pull request is ready for review with the exception of possibly adding new tests (the new code is mostly being hit by existing unit tests).

The goal of this pull request is to implement an optional randomized exponential backoff for the restart delay on task failure as per the discussion in the original issue #1201.  The new delay policy will be the default behavior, but the old constant delay functionality will still be available.  All other functionality of the restart policy will remain the same.

-Updated RestartPolicy to contain a new BackoffPolicy with a base delay, delay factor, and max delay.
-Updated CLI to allow for configuration of the backoff policy.
-Updated restart.go to use the new restart policy if applicable.  Old constant behavior will be activated if Restart.Delay != nil and Restart.Backoff == nil
-Added a new failure count to Restart.go which is reset when a task's state becomes RUNNING.  Please let us know if there is a better way to reset the failure count.

aaronlehmann commented on api/types.proto:
```diff
@@ -329,6 +329,7 @@ message RestartPolicy {
 
 	// Delay between restart attempts
 	// Note: can't use stdduration because this field needs to be nullable.
+	// Note: ingored if backoff is non null
```
typo: should be "ignored"

aaronlehmann commented on manager/orchestrator/global/global.go:
```diff
@@ -209,8 +209,11 @@ func (g *Orchestrator) handleTaskChange(ctx context.Context, t *api.Task) {
 
-	// if a task has passed running, restart it
-	if t.Status.State > api.TaskStateRunning {
+	// Notify the restart supervisor when a task runs¶
```
There's a strange character at the end of this line.

aaronlehmann commented on manager/orchestrator/global/global.go:
```diff
@@ -209,8 +209,11 @@ func (g *Orchestrator) handleTaskChange(ctx context.Context, t *api.Task) {
-	if t.Status.State > api.TaskStateRunning {
+	// Notify the restart supervisor when a task runs¶
+	if t.Status.State == api.TaskStateRunning {
+		g.restarts.Success(t)
```
I think this will work properly with health checks. If I remember correctly, the task shouldn't reach the "running" state until the health check passes.

If there is no health check, this won't work very well. The task might reach "running", briefly, and then fail.

I think this is okay, but we'd need to document that a health check is necessary for the backoff feature to work properly. It might even make sense to disallow that setup when no health check is configured.

dperny replied:
@aaronlehmann i haven't actually dug into the code enough to know if this is feasible, but what if we used logic similar to the update monitor period logic to determine a failure/success?

zapper59 replied:
Hi guys, from looking at the updater code I think that using a duration to determine when to reset the failure count would be a fairly straightforward way of doing it and would have the advantage of being independent of health checks.  I don't know enough about the code base to know which would be the "correct" solution though.
Side node:  If we were to go with a monitor duration would using a 5 second default (same as update) seem reasonable.

dperny replied:
You should use the update monitor period to determine success or failure. If you've passed the update monitor period and the task fails, it's likely a transient blip and you can go ahead and reschedule. If the update monitor period has not yet elapsed, then the failure is likely related to something about the app's bootstrapping and you should back off to wait for the cause of the failure to resolve. Same reason we use this property for updates.

aaronlehmann commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +189,87 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy
+// Returns nil on error parsing the RestartPolicy
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) *time.Duration {
```
I think it would be slightly better to return `(time.Duration, error)`.

dperny replied:
Yeah, it's considered poor form in go to use pointers as a sentinel for error conditions.

dperny commented on manager/orchestrator/restart/restart.go:
```diff
@@ -101,6 +104,7 @@ func (r *Supervisor) waitRestart(ctx context.Context, oldDelay *delayedStart, cl
 // Restart initiates a new task to replace t if appropriate under the service's
 // restart policy.
 func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Cluster, service *api.Service, t api.Task) error {
+
```
tiny nit: remove this bonus line. not a blocker to getting this merged, if you forget, but while you're cleaning up other comments just nuke this.

dperny commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +189,87 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Max); b != nil && a >= 0 {
+				backoff.Max = t.Spec.Restart.Backoff.Max
+			}
```
i'm thinking about the hilarious case in which a user deliberately sets a large base time and a negative factor in order to restart faster and faster. i'm glad you're ensuring that that does not occur.

dperny replied:
on a more serious note, there's a lot of great error checking here, but I think you should also update the service validation logic in controlapi to stop a service with these invalid parameters from even being created. (that's in addition to these checks, not instead of.)

dperny commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +189,87 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+		failures = 0
+	}
+
+	backoffDuration := base + factor*time.Duration(1<<(failures))
```
lol please don't bit shift it makes my brain hurt and i'm p sure the compiler optimizes multiply and divide by 2 anyway.

aaronlehmann replied:
@dperny: This isn't multiplication, it's computing a power of 2. Shifting is by far the simplest way to do it.

dperny replied:
oh good lord i totally missed that. i can't commutativity

dperny replied:
in that case is there any chance of an overflow?

dperny commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +189,87 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy
+// Returns nil on error parsing the RestartPolicy
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) *time.Duration {
```
This function needs a unit test.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "1201-failure-aware-scheduling" git@github.com:zapper59/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354607912
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

dperny commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +192,83 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 }
 
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) (*time.Duration, error) {
```
Still need to change this so it doesn't return a pointer. You can just `return 0, errors.New("whatever")` if there is an error. Go convention is to not look at/use the first return value if there is an error, so it doesn't really matter what the return value is, but the 0-value for the type is conventional (and the 0-value of `time.Duration` is 0).

dperny commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +192,83 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+	}
+
+	// Use floating point values to detect overflow.  math.Pow returns +Inf on overflow
+	backoffDuration := float64(base) + float64(factor)*math.Pow(2, float64(failures))
```
Thanks for changing this to `math.Pow(2, float(failures))`. I know I was wrong about the bit shift being multiplication (it wasn't) but that's part of the reason it should have been changed.

dperny commented on manager/orchestrator/replicated/restart_test.go:
```diff
@@ -803,3 +804,217 @@ func TestOrchestratorRestartWindow(t *testing.T) {
+
+	ctx := context.Background()
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
```
Like in https://github.com/docker/swarmkit/pull/2466#issuecomment-351217359, this and some other asserts are basically mandatory for the test to make any sense, so change them to `requires`.

It's tempting to use either `assert` or `require` exclusively, but sometimes you want a test to continue running when an assertion fails so that you can see what other failures occur and hopefully better understand what caused the initial test failure. But in cases like this, there's nothing meaningful to be done if the assertion fails; the whole test is useless. You can also change `require` down on 856 and 859, for example.

dperny commented on manager/orchestrator/replicated/restart_test.go:
```diff
@@ -803,3 +804,217 @@ func TestOrchestratorRestartWindow(t *testing.T) {
+	assert.Equal(t, observedTask1.ServiceAnnotations.Name, "name1")
+
+	// Check that the task has the correct BackoffPolicy values
+	Backoff1 := observedTask1.Spec.Restart.Backoff
```
style nit: don't capitalize local variable names.

dperny commented on manager/orchestrator/replicated/restart_test.go:
```diff
@@ -803,3 +804,217 @@ func TestOrchestratorRestartWindow(t *testing.T) {
+
+	// Check that the delay duration is between 0 and the calculated backoff duration
+	// assert.Equal(t, *delayP, baseTime + factorTime)
+	assert.True(t, (0 <= *delayP) && (*delayP < baseTime+factorTime))
```
If your feature doesn't work right, this test might still always succeed. For example, say the calculation fails and  `delayP == 0` always. This assertion will still pass.

That's sort of the problem with randomness, right? You can't really verify without statistical analysis.

It might be to refactor `TaskRestartDelay` such that you return the raw delay value and let the caller apply the random range. That would make it more testable. But maybe there is a better way to refactor it. I definitely don't think statistically analyzing the return value of the function in the test is the right approach tho, FWIW. 

This is not a blocker for me. Do what you feel is best.

dperny commented:
You'll need to rebase this and squash it into 1 commit before we merge. If you need help doing so, let me know, I'll give you the instructions.

Also, with this change, I think maybe it would be good to eliminate the Delay field on the service, and instead do fixed delay  by just making base = desired delay and factor = 0? WDYT?

zapper59 commented:
@dperny It seems reasonable to consolidate the two delay values.  My only concern is that removing restart.Delay would require removing the existing command line flags and refactoring all tests that reference that delay.  We designed our logic around making sure that existing taskspec's that don't reference the backoff would not have their behavior changed.  I'm not familiar enough with swarm to know how you guys handle depreciation.  I'd be open to removing the base delay and using restart.Delay instead.

dperny commented:
I think using `Delay` instead of `Base` is a good compromise, actually. You then just have to check if a factor and a max are set to compute a backoff. It would simplify your code quite a bit, too, because then you wouldn't need two separate code paths for Delay and Backoff.

anshulpundir commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +192,83 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+		backoffDuration = float64(max)
+	}
+
+	// Choose a uniformly distributed value from [0, backoffDuration).
```
Add a comment on why.

anshulpundir commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +197,83 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy.
+// It also returns whether the caller should randomized the value.
+// Important: It is the caller's responsibility to randomize the returned duration.
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) (time.Duration, bool, error) {
```
I'm curious why you decided to put the decision to randomize outside the function ?

It seems cleaner to keep it inside this function and get rid of the bool return value.

zapper59 replied:
There were concerns regarding how to test this function without having to use statistical analysis to get reliable results, and this was suggested as one of the more straightforward options.

anshulpundir commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +197,83 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy.
+// It also returns whether the caller should randomized the value.
+// Important: It is the caller's responsibility to randomize the returned duration.
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) (time.Duration, bool, error) {
```
why not use 's' for supervisor instead of 'r' ?

zapper59 replied:
We were following the convention of the rest of the file.

zapper59 commented on api/types.proto:
```diff
@@ -339,6 +340,27 @@ message RestartPolicy {
+	google.protobuf.Duration max = 3;
+
+	// Monitor indicates how long a task must run before it is considred
+	// to have started successfully. If it is restarted within Monitor of
```
It is based on when Restart is called, which happens no matter how the task finishes.  This is why we needed to use a Monitor duration in the first place as an estimation of whether the task restart was due to natural completion or an internal failure during bootstrapping.

aaronlehmann commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +197,85 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
+	}
+
+	// Use floating point values to detect overflow.  math.Pow returns +Inf on overflow
+	backoffDuration := restartDelay + factor*time.Duration(math.Pow(2, float64(failures)))
```
I think the best way to avoid overflow is to have an explicit limit on the exponent. Integer math should work fine.

aaronlehmann commented on manager/orchestrator/restart/restart.go:
```diff
@@ -190,6 +197,85 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 }
 
+// TaskRestartDelay calculates and returns the next delay based on t's RestartPolicy.
+// randomize is true if a backoff policy is being used,  false if the RestartPolicy
```
The comment refers to `randomize`, but there doesn't seem to be a variable with that name.

stevvooe commented on api/types.proto:
```diff
@@ -339,6 +341,26 @@ message RestartPolicy {
+	BackoffPolicy backoff = 5;
+}
+
+message BackoffPolicy {
```
This is not so much a policy, as it is a configuration.

zapper59 replied:
We were following the naming convention of RestartPolicy which this is a part of.

stevvooe commented on api/types.proto:
```diff
@@ -329,6 +329,8 @@ message RestartPolicy {
 
 	// Delay between restart attempts
 	// Note: can't use stdduration because this field needs to be nullable.
+	// Note: If backoff is non null then Delay will be added to the calculated
```
That is a bit surprising. It would be better to have one supplant the other.

zapper59 replied:
Delay is setup to be a base offset on top of the expo-backoff in response to an earlier comment by @dperny.  Originally we had a separate base offset which was a part of the backoff policy, and would override delay.

nishanttotla commented:
@zapper59 are you still working on this PR?

zapper59 commented:
@nishanttotla @stevvooe Sorry for the radio silence.  I think that where this pull request stands is that all code review comments have been addressed via code changes or responses to those reviews.

dperny commented:
No big deal on the radio silence from your end. We haven't really been keeping up well on our end either. I'll take another look at this tomorrow and I'll either get it merged for you or tell you what needs to change to get it through.
